client/scripts/controllers/chain/substrate/staking.ts

```diff
@@ -226,6 +257,93 @@ class SubstrateStaking implements StorageModule {
+          return commission[key];
+        });
+        // Different runtimes call for different access to stakers: old vs. new
+        const stakersCall = (api.query.staking.stakers)
+          ? api.query.staking.stakers
+          : api.query.staking.erasStakers;
+        // Different staking functions call for different function arguments: old vs. new
+        const stakersCallArgs = (account) => (api.query.staking.stakers)
+          ? account
+          : [era.toString(), account];
+        return combineLatest(
+          stakersCall.multi(currentSet.map((elt) => stakersCallArgs(elt.toString()))),
+          of(currentSet),
+          of(rewards),
+          of(commission)
+        );
+      }),
+      auditTime(100),
+      map(([exposures, accounts, rewards, commissions ] :
+        [Exposure[], AccountId[], IReward, ICommissionInfo ]) => {
+        // coins.div don't support fraction points. multiply and divide the same number will give fraction points.
+        const n = 100000;
+        const validatorRewards: ICommissionInfo = {};
+        let accountsTotalStake = new BN(0);
+
+        accounts.forEach((account, index) => {
+          const key = account.toString();
+          const exposure = exposures[index];
+          let stakeShare = 0;
+          let othersStake: BN = new BN(0);
+
+          exposure.others.forEach((indv) => {
+            othersStake = othersStake.add(indv.value.toBn());
+          });
+          othersStake = othersStake.add(exposure.own.toBn());
+          if (othersStake.gt(new BN(0))) {
+            stakeShare = exposure.own.toBn().muln(n).div(othersStake).toNumber() / n;
+          }
+
+          const totalStake = exposure?.total.toBn() || new BN(0);
+          let totalReward = rewards.validators[key] || 0;
+
+          // if rewards have rewards for chain id, split with every validator equally.
+          if (rewards.validators[this._app.chain.id])
+            totalReward += rewards.validators[this._app.chain.id] / accounts.length;
+
+          accountsTotalStake = accountsTotalStake.add(totalStake);
+
+          const commissionPercent: number = (commissions[key] || 0) / 100;
+
+          let rewardEarn = totalReward * commissionPercent;
+          const othersReward = totalReward - rewardEarn;
+          rewardEarn += stakeShare * othersReward;
+          let percentage = 0;
+
+          if (totalStake.gt(new BN(0))) {
+            const rewardBN = (this._app.chain as Substrate).chain.coins(rewardEarn);
+            const totalStakeBN = (this._app.chain as Substrate).chain.coins(totalStake);
+            // Number can only safely store up to 53 bits for toNumber function.
+            percentage = +rewardBN.muln(n).div(totalStakeBN).toString() / n;
+          }
+          // rewards.daysDiff
+          // number of days between last reward record and latest reward record through events to ChainEvents
+          validatorRewards[key] = ((percentage / (rewards.daysDiff || 1)) * 365) / 100;
```

Comment:
What does this math do/what does this value mean?

Reply:
calculating APR - Annual Percentage Rate for a validator.

( ( ( total earnings / total stake ) / total days of rewards recorded) * one year) / 100

Reply:
The problem with this approach is it does not actually line up so nicely into days, since Kusama validators can claim rewards over many different eras and I'm not sure we're tracking this data. The events only show us the amount rewarded and not how many eras it's composed of.